### PR TITLE
Support for autoimports

### PR DESCRIPTION
**Pull request by PolinB**

### CODE REVIEW

File: src/main/kotlin/com/compiler/server/compiler/components/SuggestionProvider.kt [hunk 1]
```diff
@@ -202,14 +215,14 @@ class CompletionProvider(private val errorAnalyzer: ErrorAnalyzer) {
...
-
   private fun formatName(
     builder: String,
     symbols: Int
   ) = if (builder.length > symbols) builder.substring(0, symbols) + "..." else builder
 
   private fun keywordsCompletionVariants(keywords: TokenSet, prefix: String) = keywords.types.mapNotNull {
-    if (it is KtKeywordToken && it.value.startsWith(prefix)) Completion(it.value, it.value, "", "") else null
+    if (it is KtKeywordToken && it.value.startsWith(prefix))
+      Completion(it.value, it.value, "", "", "") else null
```
Comment (AlexanderPrendota) [inline src/main/kotlin/com/compiler/server/compiler/components/SuggestionProvider.kt:225]: default values?

File: src/main/kotlin/com/compiler/server/compiler/components/SuggestionProvider.kt [hunk 2]
```diff
@@ -119,7 +132,8 @@ class CompletionProvider(private val errorAnalyzer: ErrorAnalyzer) {
       Completion(
         text = completionText,
         displayText = fullName,
-        tail = formatName(fullName, NUMBER_OF_CHAR_IN_TAIL),
+        tail = formatName(type, NUMBER_OF_CHAR_IN_TAIL),
+        import = "",
```
Comment (AlexanderPrendota) [inline src/main/kotlin/com/compiler/server/compiler/components/SuggestionProvider.kt:136]: Why import is an empty string? AFAIK it's a list of strings

File: src/main/kotlin/com/compiler/server/compiler/components/SuggestionProvider.kt [hunk 3]
```diff
@@ -61,15 +69,21 @@ class CompletionProvider(private val errorAnalyzer: ErrorAnalyzer) {
       val descriptorInfo = descriptorsFrom(this, element, isJs, coreEnvironment)
       val prefix = (if (descriptorInfo.isTipsManagerCompletion) element.text else element.parent.text)
         .substringBefore("IntellijIdeaRulezzz").let { if (it.endsWith(".")) "" else it }
+      val importCompletionVariants = try {
```
Comment (AlexanderPrendota) [inline src/main/kotlin/com/compiler/server/compiler/components/SuggestionProvider.kt:72]: runCatching

File: src/main/kotlin/com/compiler/server/compiler/components/SuggestionProvider.kt [hunk 4]
```diff
@@ -28,10 +32,12 @@ import org.jetbrains.kotlin.resolve.scopes.MemberScope
 import org.jetbrains.kotlin.types.asFlexibleType
 import org.jetbrains.kotlin.types.isFlexible
 import org.springframework.stereotype.Component
-
+import java.io.File
 
 @Component
-class CompletionProvider(private val errorAnalyzer: ErrorAnalyzer) {
+class SuggestionProvider(
+  private val errorAnalyzer: ErrorAnalyzer
```
Comment (AlexanderPrendota) [inline src/main/kotlin/com/compiler/server/compiler/components/SuggestionProvider.kt:39]: refactor only in a separate branch

File: indexation/src/main/kotlin/Main.kt [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    File(outputPath).writeText("")
+
+    val mapper = jacksonObjectMapper()
+    File(outputPath).appendText(mapper.writerWithDefaultPrettyPrinter()
+      .writeValueAsString(getAllVariants(classLoader, files)))
+//    File(outputPath).appendText(mapper.writeValueAsString(getAllVariants(classLoader, files)))
+  }
+
+  @JvmStatic
+  // First argument is path to folder with jars
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:223]: /**

**/

File: indexation/src/main/kotlin/Main.kt [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+    val file = File(directoryPath)
+    val files = file.listFiles().toList()
+    val classPathUrls = initClasspath(directoryPath)
+    val classLoader = URLClassLoader.newInstance(classPathUrls.toTypedArray())
+    File(outputPath).writeText("")
+
+    val mapper = jacksonObjectMapper()
+    File(outputPath).appendText(mapper.writerWithDefaultPrettyPrinter()
+      .writeValueAsString(getAllVariants(classLoader, files)))
+//    File(outputPath).appendText(mapper.writeValueAsString(getAllVariants(classLoader, files)))
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:219]: do u need that?

File: indexation/src/main/kotlin/Main.kt [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+  }
+
+  private fun createJsonWithIndexes(directoryPath: String, outputPath: String) {
+    val file = File(directoryPath)
+    val files = file.listFiles().toList()
+    val classPathUrls = initClasspath(directoryPath)
+    val classLoader = URLClassLoader.newInstance(classPathUrls.toTypedArray())
+    File(outputPath).writeText("")
+
+    val mapper = jacksonObjectMapper()
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:216]: global constant

File: indexation/src/main/kotlin/Main.kt [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+      Modifier.isStatic(method.modifiers))
+      importInfoByMethodAndParent(
+        method.name,
+        method.parameters.joinToString { it.type.name },
+        method.returnType.simpleName,
+        clazz
+      )
+    else null
+
+  private fun importInfoByMethodAndParent(methodName: String,
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:187]: fomatting

File: indexation/src/main/kotlin/Main.kt [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+    do {
+      range = regex.find(type)?.groups?.get(1)?.range
+      type = if (range != null) type.removeRange(range) else type
+    } while (range != null)
+    return type
+  }
+
+  private fun importInfoFromJavaMethod(method: Method, clazz: Class<*>): ImportInfo? =
+    if (Modifier.isPublic(method.modifiers) &&
+      Modifier.isStatic(method.modifiers))
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:178]: fomatting

File: indexation/src/main/kotlin/Main.kt [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+      importInfoByMethodAndParent(
+        method.name,
+        method.parameters.joinToString { it.type.name },
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:181]: named arguments

File: indexation/src/main/kotlin/Main.kt [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+          kotlinTypeToType(it.type)
+        },
+        kotlinTypeToType(kotlinFunction.returnType),
+        clazz)
+    } else importInfoFromJavaMethod(method, clazz)
+  }
+
+  private fun kotlinTypeToType(kotlinType: KType): String {
+    var type = kotlinType.toString()
+    val regex = Regex("(kotlin\\.)([A-Z])")
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:167]: move regex to the constant and create a comment with description

File: indexation/src/main/kotlin/Main.kt [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+    }
+    clazz.declaredMethods.map { method ->
+      val importInfo = importInfoFromFunction(method, clazz)
+      if (importInfo != null) allClasses.add(importInfo)
+    }
+    return allClasses
+  }
+
+  private fun importInfoFromFunction(method: Method, clazz: Class<*>): ImportInfo? {
+    var kotlinFunction: KFunction<*>? = null
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:145]: val kotlinFunction = try {}...

File: indexation/src/main/kotlin/Main.kt [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+    try {
+      kotlinFunction = method.kotlinFunction
+    } catch (exception: NoSuchElementException) {
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:148]: only 4? =) maybe runCatching?

File: indexation/src/main/kotlin/Main.kt [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+    } catch (exception: UnsupportedOperationException) {
+    } catch (error: KotlinReflectionInternalError) {
+    } catch (error: IncompatibleClassChangeError) {
+    }
+    return if (kotlinFunction != null
+      && kotlinFunction.visibility == KVisibility.PUBLIC) {
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:154]: formatting

File: indexation/src/main/kotlin/Main.kt [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+      importInfoByMethodAndParent(
+        kotlinFunction.name,
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:156]: named arguments

File: indexation/src/main/kotlin/Main.kt [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+        classPath.add(file)
+      }
+    }
+    return classPath.mapNotNull { it.toURI().toURL() }
+  }
+
+  private fun getVariantsForZip(classLoader: URLClassLoader, file: File): List<ImportInfo> {
+    val jarFile = JarFile(file)
+    val allSuggests = hashSetOf<ImportInfo>()
+    jarFile.entries().toList().map { entry ->
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:111]: why you need .map over here? as I see you don't need a return value. use a plain loop instead

File: indexation/src/main/kotlin/Main.kt [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+      if (!entry.isDirectory
+        && entry.name.endsWith(CLASS_EXTENSION)
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:113]: formatting

File: indexation/src/main/kotlin/Main.kt [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+import java.net.URLClassLoader
+import java.util.jar.JarFile
+import kotlin.reflect.KFunction
+import kotlin.reflect.KType
+import kotlin.reflect.KVisibility
+import kotlin.reflect.jvm.internal.KotlinReflectionInternalError
+import kotlin.reflect.jvm.kotlinFunction
+
+object Main {
+  private data class ImportInfo(val importName: String,
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:18]: formatting

File: src/main/kotlin/com/compiler/server/model/ImportInfo.kt [hunk 19]
```diff
@@ -0,0 +1,7 @@
+package com.compiler.server.model
+
+data class ImportInfo (val importName: String,
```
[the remark this answers is not in the document]
Comment (PolinB) [reply src/main/kotlin/com/compiler/server/model/ImportInfo.kt]: done

File: src/main/kotlin/com/compiler/server/model/Completion.kt [hunk 20]
```diff
@@ -1,3 +1,9 @@
 package com.compiler.server.model
 
-data class Completion(val text: String, val displayText: String, val tail: String, val icon: String)
\ No newline at end of file
+data class Completion(
+  val text: String,
+  val displayText: String,
+  val tail: String = "",
+  val import: String = "",
```
Comment (AlexanderPrendota) [inline src/main/kotlin/com/compiler/server/model/Completion.kt:7]: maybe null?

File: src/main/kotlin/com/compiler/server/compiler/components/CompletionProvider.kt [hunk 21]
```diff
@@ -262,4 +274,40 @@ class CompletionProvider(private val errorAnalyzer: ErrorAnalyzer) {
...
+
+  fun checkUnresolvedReferences(errors: Map<String, List<ErrorDescriptor>>): Map<String, List<ErrorDescriptor>> =
+    errors.mapValues { (_, errorDescriptors) ->
+      return@mapValues errorDescriptors.map { errorDescriptor ->
+        checkErrorDescriptor(errorDescriptor)
+      }
+    }
+
+  private fun checkErrorDescriptor(errorDescriptor: ErrorDescriptor): ErrorDescriptor {
+    if (errorDescriptor.message.startsWith(UNRESOLVED_REFERENCE_MESSAGE_PREFIX)) {
```
Comment (AlexanderPrendota) [inline src/main/kotlin/com/compiler/server/compiler/components/CompletionProvider.kt:294]: if (!errorDescriptor.message.startsWith(UNRESOLVED_REFERENCE_MESSAGE_PREFIX)) return errorDescriptor

File: indexation/src/main/kotlin/Main.kt [hunk 22]
```diff
@@ -0,0 +1,180 @@
...
+        }
+      }
+    }
+    return allSuggests.toList()
+  }
+
+  private fun allFunctionsFromClass(clazz: Class<*>): HashSet<ImportInfo> =
+    (clazz.methods + clazz.declaredMethods).mapNotNull { method ->
+      importInfoFromFunction(method, clazz)
+    }.toHashSet()
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:105]: Why you need HashSet everywhere?

File: indexation/src/main/kotlin/Main.kt [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+    File(outputPath).writeText("")
+    File(outputPath).appendText(jacksonObjectMapper().writeValueAsString(getAllVariants(classLoader, files)))
+  }
+
+  @JvmStatic
+  /**
+   * First argument is path to folder with jars
+   * Second argument is path to output file
+   **/
+  fun main(args: Array<String>) {
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:175]: move public function to the top of the class

File: indexation/src/main/kotlin/Main.kt [hunk 24]
```diff
@@ -0,0 +1,180 @@
...
+    }.map { file ->
+      getVariantsForZip(classLoader, file)
+    }.flatten()
+
+  private fun createJsonWithIndexes(directoryPath: String, outputPath: String) {
+    val files = File(directoryPath).listFiles().toList()
+    val classPathUrls = initClasspath(directoryPath)
+    val classLoader = URLClassLoader.newInstance(classPathUrls.toTypedArray())
+
+    File(outputPath).writeText("")
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:166]: why?

File: src/main/kotlin/com/compiler/server/model/Completion.kt [hunk 25]
```diff
@@ -1,3 +1,9 @@
 package com.compiler.server.model
 
-data class Completion(val text: String, val displayText: String, val tail: String, val icon: String)
\ No newline at end of file
+data class Completion(
+  val text: String,
+  val displayText: String,
+  val tail: String = "",
+  val import: String? = null,
+  val icon: String = ""
```
Comment (AlexanderPrendota) [inline src/main/kotlin/com/compiler/server/model/Completion.kt:8]: why empty?

File: indexation/src/main/kotlin/Main.kt [hunk 26]
```diff
@@ -0,0 +1,186 @@
...
+            allClassesFromJavaClass(clazz)
+          }
+          val functions = allFunctionsFromClass(clazz)
+          return@map classes + functions
+        } else emptyList<ImportInfo>()
+      } else emptyList()
+    }.flatten().distinct()
+
+  private fun allFunctionsFromClass(clazz: Class<*>): List<ImportInfo> =
+    (clazz.methods + clazz.declaredMethods).distinct().mapNotNull { method ->
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:116]: 2 distinct?

File: indexation/src/main/kotlin/Main.kt [hunk 27]
```diff
@@ -0,0 +1,186 @@
...
+        }
+      } else {
+        classPath.add(file)
+      }
+    }
+    return classPath.mapNotNull { it.toURI().toURL() }
+  }
+
+  private fun getVariantsForZip(classLoader: URLClassLoader, file: File): List<ImportInfo> =
+    JarFile(file).entries().toList().map { entry ->
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:98]: flatMap?

File: indexation/src/main/kotlin/Main.kt [hunk 28]
```diff
@@ -0,0 +1,186 @@
...
+        val fullName = name.replace(File.separator, ".")
+        if (fullName != MODULE_INFO_NAME) {
+          val clazz = classLoader.loadClass(fullName) ?: return@map emptyList<ImportInfo>()
+          val classes = if (clazz.isKotlinClass()) {
+            allClassesFromKotlinClass(clazz)
+          } else {
+            allClassesFromJavaClass(clazz)
+          }
+          val functions = allFunctionsFromClass(clazz)
+          return@map classes + functions
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:110]: return@map classes + functions
        } else emptyList<ImportInfo>()
      } else emptyList()
    }.flatten().distinct()


possible to simplify? =)

File: common/src/main/kotlin/model/Completion.kt [hunk 29]
```diff
@@ -0,0 +1,9 @@
+package common.model
+
+data class Completion(
```
Comment (AlexanderPrendota) [inline common/src/main/kotlin/model/Completion.kt:3]: @JsonInclude(JsonInclude.Include.NON_NULL)

File: indexation/src/main/kotlin/Main.kt [hunk 30]
```diff
@@ -0,0 +1,189 @@
...
+      )
+    }
+
+  private fun allClassesFromKotlinClass(clazz: Class<*>): List<ImportInfo> =
+    runCatching {
+      val kotlinClass = clazz.kotlin
+      val result = clazz.kotlin.nestedClasses.filter {
+        it.visibility == KVisibility.PUBLIC
+      }.map {
+        val canonicalName = it.qualifiedName ?: ""
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:58]: why need an empty string? why not null?

File: indexation/src/main/kotlin/Main.kt [hunk 31]
```diff
@@ -0,0 +1,189 @@
...
+      } else {
+        classPath.add(file)
+      }
+    }
+    return classPath.mapNotNull { it.toURI().toURL() }
+  }
+
+  private fun getVariantsForZip(classLoader: URLClassLoader, file: File): List<ImportInfo> =
+    JarFile(file).entries().toList().flatMap { entry ->
+      if (!entry.isDirectory && entry.name.endsWith(CLASS_EXTENSION)) {
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:100]: maybe just use .filter {} ?

File: indexation/src/main/kotlin/Main.kt [hunk 32]
```diff
@@ -0,0 +1,189 @@
...
+      fullName = className,
+      returnType = returnType,
+      icon = METHOD_ICON
+    )
+  }
+
+  private fun getAllVariants(classLoader: URLClassLoader, files: List<File>): List<ImportInfo> =
+    files.filter { jarFile ->
+      jarFile.name.split(File.separator).last() != EXECUTORS_JAR_NAME
+    }.map { file ->
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:179]: use default it like .map { getVariantsForZip(c, it )}

File: src/main/kotlin/com/compiler/server/compiler/components/CompletionProvider.kt [hunk 33]
```diff
@@ -56,20 +63,26 @@ class CompletionProvider(private val errorAnalyzer: ErrorAnalyzer) {
     character: Int,
     isJs: Boolean,
     coreEnvironment: KotlinCoreEnvironment
-  ) = with(file.insert("IntellijIdeaRulezzz ", line, character)) {
+  ) : List<Completion> = with(file.insert("IntellijIdeaRulezzz ", line, character)) {
```
Comment (AlexanderPrendota) [inline src/main/kotlin/com/compiler/server/compiler/components/CompletionProvider.kt:66]: IntellijIdeaRulezzz - constant

File: src/main/kotlin/com/compiler/server/compiler/components/CompletionProvider.kt [hunk 34]
```diff
@@ -262,4 +273,12 @@ class CompletionProvider(private val errorAnalyzer: ErrorAnalyzer) {
...
   }
-}
\ No newline at end of file
+
+  private fun readIndexesFromJson(): List<ImportInfo> =
+    jacksonObjectMapper().readValue(File(indexesFileName).readText())
+
+  private fun getClassesByName(name: String) =
+    readIndexesFromJson().filter { variant ->
+      variant.shortName == name
```
Comment (AlexanderPrendota) [inline src/main/kotlin/com/compiler/server/compiler/components/CompletionProvider.kt:282]: it

File: indexation/src/main/kotlin/Main.kt [hunk 35]
```diff
@@ -0,0 +1,187 @@
...
+      )
+    }
+
+  private fun allClassesFromKotlinClass(clazz: Class<*>): List<ImportInfo> =
+    runCatching {
+      val kotlinClass = clazz.kotlin
+      val result = clazz.kotlin.nestedClasses.filter {
+        it.visibility == KVisibility.PUBLIC
+      }.mapNotNull {
+        val canonicalName = it.qualifiedName ?: return@mapNotNull null
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:58]: maybe `.filter { x != null && y != null }`

File: indexation/src/main/kotlin/Main.kt [hunk 36]
```diff
@@ -0,0 +1,187 @@
...
+        fullName = simpleName,
+        returnType = simpleName,
+        icon = CLASS_ICON
+      )
+    }
+
+  private fun allClassesFromKotlinClass(clazz: Class<*>): List<ImportInfo> =
+    runCatching {
+      val kotlinClass = clazz.kotlin
+      val result = clazz.kotlin.nestedClasses.filter {
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:55]: why name result? maybe `nestedClassesImports`

File: indexation/src/main/kotlin/Main.kt [hunk 37]
```diff
@@ -0,0 +1,187 @@
...
+        it.visibility == KVisibility.PUBLIC
+      }.mapNotNull {
+        val canonicalName = it.qualifiedName ?: return@mapNotNull null
+        val simpleName = it.simpleName ?: return@mapNotNull null
+        ImportInfo(canonicalName, simpleName, simpleName, simpleName, CLASS_ICON)
+      }
+      val classInfo = if (kotlinClass.visibility == KVisibility.PUBLIC) {
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:62]: classInfo -> classInfoImports?

File: indexation/src/main/kotlin/Main.kt [hunk 38]
```diff
@@ -0,0 +1,187 @@
...
+        val canonicalName = kotlinClass.qualifiedName ?: ""
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:63]: why an empty string? maybe null?

File: indexation/src/main/kotlin/Main.kt [hunk 39]
```diff
@@ -115,20 +110,26 @@ object Main {
   private fun allFunctionsFromClass(clazz: Class<*>): List<ImportInfo> =
     (clazz.methods + clazz.declaredMethods).distinct().mapNotNull { method ->
       importInfoFromFunction(method, clazz)
+    }.filter {
+      it.shortName != EQUALS_METHOD && it.shortName != HASH_CODE_METHOD && it.shortName != TO_STRING_METHOD
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/Main.kt:114]: `it !in set` ? =)

File: src/main/kotlin/com/compiler/server/compiler/components/CompletionProvider.kt [hunk 40]
```diff
@@ -52,6 +58,7 @@ class CompletionProvider(
   private val NUMBER_OF_CHAR_IN_COMPLETION_NAME = 40
   private val NAME_FILTER = { name: Name -> !name.isSpecial }
   private val COMPLETION_SUFFIX = "IntellijIdeaRulezzz"
+  private var ALL_INDEXES: List<ImportInfo>? = null
```
Comment (AlexanderPrendota) [inline src/main/kotlin/com/compiler/server/compiler/components/CompletionProvider.kt:61]: or better:  val ALL_INDEXES = kotlin.runCatching { readIndexesFromJson() }.getOrNull() =)

File: src/main/kotlin/com/compiler/server/compiler/components/CompletionProvider.kt [hunk 41]
```diff
@@ -27,11 +30,21 @@ import org.jetbrains.kotlin.resolve.scopes.DescriptorKindFilter
...
 @Component
-class CompletionProvider(private val errorAnalyzer: ErrorAnalyzer) {
+class CompletionProvider(
+  private val errorAnalyzer: ErrorAnalyzer,
+  @Value("\${indexes.file}") private val indexesFileName: String
+) {
+
+  @PostConstruct
+  private fun postConstruct() {
+    ALL_INDEXES = kotlin.runCatching { readIndexesFromJson() }.getOrNull()
```
Comment (AlexanderPrendota) [inline src/main/kotlin/com/compiler/server/compiler/components/CompletionProvider.kt:46]: if null then log.warn("bla bla server started w/o indexes..."

File: src/main/kotlin/com/compiler/server/compiler/components/CompletionProvider.kt [hunk 42]
```diff
@@ -54,12 +50,21 @@ class CompletionProvider(
...
   private val NUMBER_OF_CHAR_IN_COMPLETION_NAME = 40
   private val NAME_FILTER = { name: Name -> !name.isSpecial }
   private val COMPLETION_SUFFIX = "IntellijIdeaRulezzz"
   private var ALL_INDEXES: List<ImportInfo>? = null
 
+  @PostConstruct
+  private fun initIndexes() {
+    ALL_INDEXES = kotlin.runCatching { readIndexesFromJson() }.getOrNull()
+    if (ALL_INDEXES == null) {
+      log.warn("Exception in getting indexes. Server started without auto imports.")
```
Comment (AlexanderPrendota) [inline src/main/kotlin/com/compiler/server/compiler/components/CompletionProvider.kt:64]: Server started without auto imports. - will be enough

File: common/src/main/kotlin/model/ImportInfo.kt [hunk 43]
```diff
@@ -0,0 +1,18 @@
+package common.model
+
+data class ImportInfo(
```
Comment (zarechenskiy) [inline common/src/main/kotlin/model/ImportInfo.kt:3]: Minor but I'd put this class and `Completion` into one file, Kotlin allows doing this :)

File: indexation/src/main/kotlin/DisplayingTypes.kt [hunk 44]
```diff
@@ -0,0 +1,44 @@
+package indexation
+
+import org.jetbrains.kotlin.builtins.PrimitiveType
+import org.jetbrains.kotlin.builtins.jvm.JavaToKotlinClassMap
+import org.jetbrains.kotlin.name.FqName
+import kotlin.reflect.KType
+
+private const val KOTLIN_TYPE_PREFIX = "(kotlin\\.)([A-Z])" // prefix for simple kotlin type, like Double, Any...
+
+private val primitiveToPrimitiveType: Map<String, PrimitiveType> = mapOf(
```
Comment (zarechenskiy) [inline indexation/src/main/kotlin/DisplayingTypes.kt:10]: Minor: it seems `Map<String, PrimitiveType>` can be omitted

File: indexation/src/main/kotlin/DisplayingTypes.kt [hunk 45]
```diff
@@ -0,0 +1,44 @@
+package indexation
+
+import org.jetbrains.kotlin.builtins.PrimitiveType
+import org.jetbrains.kotlin.builtins.jvm.JavaToKotlinClassMap
+import org.jetbrains.kotlin.name.FqName
+import kotlin.reflect.KType
+
+private const val KOTLIN_TYPE_PREFIX = "(kotlin\\.)([A-Z])" // prefix for simple kotlin type, like Double, Any...
```
Comment (zarechenskiy) [inline indexation/src/main/kotlin/DisplayingTypes.kt:8]: Always save only `Regex` objects: `Regex(""(kotlin\\.)([A-Z])")`, not strings as it'll lead to performance problems later. Constructor of Regex performs `Pattern.compile(pattern)` which isn't fast

File: indexation/src/main/kotlin/DisplayingTypes.kt [hunk 46]
```diff
@@ -0,0 +1,44 @@
...
+    return primitiveToPrimitiveType[type.simpleName]?.typeName?.asString() ?: type.simpleName
+  } else if (type.isArray) {
+    val componentType = type.componentType
+    return if (componentType.isPrimitive) {
+      primitiveToPrimitiveType[componentType.simpleName]?.arrayTypeName?.asString() ?: type.simpleName
+    } else {
+      return "Array<${javaTypeToKotlin(componentType)}>"
+    }
+  }
+  return JavaToKotlinClassMap.mapJavaToKotlin(FqName(type.canonicalName))?.shortClassName?.asString() ?: type.simpleName
```
Comment (zarechenskiy) [inline indexation/src/main/kotlin/DisplayingTypes.kt:32]: Currently, this code is written in Java-style, let's rewrite it to something:
```
return when {
    type.isPrimitive -> ... 
    type.isArray ->

}
```

This way we'll avoid duplicates of `if`/`else if` and several `return`s

File: indexation/src/main/kotlin/Main.kt [hunk 47]
```diff
@@ -0,0 +1,171 @@
...
+import kotlin.reflect.KVisibility
+import kotlin.reflect.jvm.kotlinFunction
+
+object Main {
+  @JvmStatic
+    /**
+     * First argument is path to folder with jars
+     * Second argument is path to output file
+     **/
+  fun main(args: Array<String>) {
```
Comment (zarechenskiy) [inline indexation/src/main/kotlin/Main.kt:21]: It's better to put comments after annotations like `@JvmStatic`

File: indexation/src/main/kotlin/Main.kt [hunk 48]
```diff
@@ -0,0 +1,171 @@
...
+import java.io.File
+import java.lang.reflect.Method
+import java.lang.reflect.Modifier
+import java.net.URL
+import java.net.URLClassLoader
+import java.util.jar.JarFile
+import kotlin.reflect.KVisibility
+import kotlin.reflect.jvm.kotlinFunction
+
+object Main {
```
Comment (zarechenskiy) [inline indexation/src/main/kotlin/Main.kt:15]: What is the reason for having this `object`?

**Review by zarechenskiy — APPROVED** [review]
Looks good!

File: indexation/src/main/kotlin/Main.kt (hunk 48 above)
Comment (PolinB) [reply indexation/src/main/kotlin/Main.kt:15]: It's for run application from Gradle with parameters
`application {
    mainClassName = "indexation.Main"
}`
If I remove `object` I get an error: `Error: Could not find or load main class indexation.Main`

**Pull request merged**